### PR TITLE
Data_Engine: Adding create methods and first query methods for Table class

### LINKED ISSUE #1285: Data_Engine: Add create and Query methods for Table
#### Description:

Add methods for generating table objects added in https://github.com/BHoM/BHoM/issues/593.

First few query methods to get data from the table also to be added.

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/595
   
### Issues addressed by this PR

Closes #1285 

Adding create methods and first Query methods for Tables.

Please note that the DataTable can not be serialised to Bson by default. I have code ready for this, but will need to be aligned with the work done by @adecler in https://github.com/BHoM/BHoM_Engine/pull/1270

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Data_Engine/Issue_1289_CreateTables?csf=1&e=KcLoUo

### Changelog

- Adding Create method for table class. One from `List<IBHoMObject>` and one from 2D table with header values.
- Query method for getting table rows matching exact values in table
- Query method for getting table rows matching values above set axis-values

### Additional comments

@Martian42

### CODE REVIEW

File: Data_Engine/Create/Table.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            if (ignoreGuid && dict.ContainsKey("BHoM_Guid"))
+                dict.Remove("BHoM_Guid");
+            if (ignoreTags && dict.ContainsKey("Tags"))
+                dict.Remove("Tags");
+
+        }
+
+        /***************************************************/
+
+        [Description("")]
```
Comment (epignatelli) [inline Data_Engine/Create/Table.cs]: Are these going to be filled? If not, we can remove them?

File: Data_Engine/Query/Axes.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace BH.Engine.Data
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("")]
```
Comment (epignatelli) [inline Data_Engine/Query/Axes.cs]: As above

File: Data_Engine/Create/Table.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        [Input("", "")]
+        [Output("", "")]
+        public static Table Table<T>(string name, string axis1Name, List<IComparable> axis1Values, string axis2Name, List<IComparable> axis2Values, List<List<T>> values, string valuesName = "Values")
```
Comment (epignatelli) [inline Data_Engine/Create/Table.cs]: What is an `IComparable` exactly? Can we input it from the UI side of things?
Not commenting all the rest, it's in other places as well.

File: Data_Engine/Create/Table.cs (hunk 1 above)
Comment (IsakNaslundBh) [reply Data_Engine/Create/Table.cs]: Thanks @epignatelli , yes, need to get all of these filled in.

**Comment by IsakNaslundBh** [comment]
@adecler , would not mind a review from you on this, especially on the Serialiser_Engine part.

File: Data_Engine/Create/Table.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            Dictionary<string, object> properties = first.PropertyDictionary();
+
+            List<DataColumn> columns = new List<DataColumn>();
+
+            properties.CleanUnwantedProperties(ignoreName, ignoreGuid, ignoreTags);
+
+            foreach (var kvp in properties)
+            {
+                
+                columns.Add(new DataColumn(kvp.Key, kvp.Value.GetType()));
```
Comment (adecler) [inline Data_Engine/Create/Table.cs]: This needs a null check:

image

**Comment by IsakNaslundBh** [comment]
Pinging this!

Would anyone please be able to have a look at this PR so we can get it merged. Thanks!

**Comment by FraserGreenroyd** [comment]
Sorry @IsakNaslundBh yes I'll try and get on this today!

**Comment by FraserGreenroyd** [comment]
> No pipelines are associated with this pull request.

Good to know, thanks bot 🤦‍♂

File: Data_Engine/Query/Axes.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Gets the names of all the axes in the table")]
+        [Input("table", "The table to extract the axes from")]
+        [Output("axes", "The table axes names")]
+        public static List<string> Axes(this Table table)
+        {
+            var columnNames = new List<string>();
```
Comment (FraserGreenroyd) [inline Data_Engine/Query/Axes.cs]: Could we not simplify this with a Linq query instead?
`return table.Data.Columns.Select(x => x.ColumnName).ToList();`

File: Data_Engine/Create/Table.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            DataTable table = new DataTable();
+
+            DataColumn[] cols =
+                {
+                new DataColumn(axis1Name, axis1Values.First().GetType()),
+                new DataColumn(axis2Name, axis2Values.First().GetType()),
+                new DataColumn(valuesName, typeof(T))
+            };
+
```
Comment (FraserGreenroyd) [inline Data_Engine/Create/Table.cs]: Excess whitespace

File: Data_Engine/Create/Table.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+            table.Columns.AddRange(cols);
+
```
Comment (FraserGreenroyd) [inline Data_Engine/Create/Table.cs]: Excess whitespace

File: Data_Engine/Create/Table.cs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                    DataRow row = table.NewRow();
+                    row[axis1Name] = axis1Values[i];
+                    row[axis2Name] = axis2Values[j];
+                    row[valuesName] = values[i][j];
+                    table.Rows.Add(row);
+                }
+            }
+
+            return new Table() { Data = table, Name = name };
+
```
Comment (FraserGreenroyd) [inline Data_Engine/Create/Table.cs]: Excess whitespace

File: Data_Engine/Create/Table.cs [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                    row[axis1Name] = axis1Values[i];
+                    row[axis2Name] = axis2Values[j];
+                    row[valuesName] = values[i][j];
+                    table.Rows.Add(row);
+                }
+            }
+
+            return new Table() { Data = table, Name = name };
+            
+
```
Comment (FraserGreenroyd) [inline Data_Engine/Create/Table.cs]: Excess whitespace

File: Data_Engine/Create/Table.cs [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            foreach (var prop in obj.GetType().GetProperties())
+            {
+                if (!prop.CanRead || prop.GetMethod.GetParameters().Count() > 0) continue;
+
+                dic[prop.Name] = prop.PropertyType;
+            }
+            return dic;
+        }
+
```
Comment (FraserGreenroyd) [inline Data_Engine/Create/Table.cs]: Excess whitespace

**Comment by al-fisher** [comment]
> No pipelines are associated with this pull request.

>Good to know, thanks bot 🤦‍♂


@FraserGreenroyd it is the beginning of the end! 🤖 
😆

File: Data_Engine/Create/Table.cs [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+                dic["Name"] = typeof(string);
+
+            if (!dic.ContainsKey("BHoM_Guid"))
+                dic["BHoM_Guid"] = typeof(Guid);
+
+            if (!dic.ContainsKey("Tags"))
+                dic["Tags"] = obj.Tags.GetType();
+
+            return dic;
+
```
Comment (FraserGreenroyd) [inline Data_Engine/Create/Table.cs]: Excess whitespace

**Comment by FraserGreenroyd** [comment]
> > No pipelines are associated with this pull request.
> 
> > Good to know, thanks bot 🤦‍♂
> 
> @FraserGreenroyd it is the beginning of the end! 🤖
> 😆

Skynet here we come... time to build a bunker! *tinfoil hat firmly placed*

**Comment by FraserGreenroyd** [comment]
Test script worked with some modifications (I think the GH file was slightly out of date to the latest code, but replacing the components worked fine for me), so am happy to approve this PR when the above code comments are addressed @IsakNaslundBh :smile:

File: Data_Engine/Query/Axes.cs (hunk 5 above)
Comment (IsakNaslundBh) [reply Data_Engine/Query/Axes.cs]: Tried that at first, but can not use Linq on table.Data.Columns for whatever reason. Not sure why atm.

Could do:  `            return table.Data.Columns.Cast<DataColumn>().Select(x => x.ColumnName).ToList();` if you think that is cleaner?

**Comment by IsakNaslundBh** [comment]
Thanks @FraserGreenroyd 🚀 

Have fixed your white space comments. Also added a method to get all data out from the table. Added that method to the testscript as well and updated the outdated components.

**Pull request merged**